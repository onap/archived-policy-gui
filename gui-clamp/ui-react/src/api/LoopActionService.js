/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopActionService{

	static performAction(cl_name, uiAction) {
		console.info("LoopActionService perform action: " + uiAction + " closedloopName=" + cl_name);
		const svcAction = uiAction.toLowerCase();
 		return fetch(window.location.pathname + "restservices/clds/v2/loop/" + svcAction + "/" + cl_name, {
 				method: 'PUT',
 				credentials: 'same-origin'
 			})
 		.then(function (response) {
 			if (response.ok) {
 				return response.json();
 			} else {
 				return Promise.reject("Perform action failed with code:" + response.status);
 			}
 		})
 		.then(function (data) {
			console.info("Action Successful: " + uiAction);
 			return data;
 		})
 		.catch(function(error) {
 			console.info("Action Failure: " + uiAction);
 			return Promise.reject(error);
 		});
 	}


	static refreshStatus(cl_name) {
		console.info("Refresh the status for closedloopName=" + cl_name);

		return fetch(window.location.pathname + "restservices/clds/v2/loop/getstatus/" + cl_name, {
			method: 'GET',
			credentials: 'same-origin'
		})
		.then(function (response) {
			if (response.ok) {
				return response.json();
			} else {
				return Promise.reject("Refresh status failed with code:" + response.status);
			}
		})
		.then(function (data) {
			console.info ("Refresh status Successful");
			return data;
		})
		.catch(function(error) {
			console.info ("Refresh status failed:", error);
			return Promise.reject(error);
		});
 	}
}
