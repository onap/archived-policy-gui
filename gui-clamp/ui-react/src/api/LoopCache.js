/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopCache {
	loopJsonCache;

	constructor(loopJson) {
		this.loopJsonCache=loopJson;
	}

	updateMicroServiceProperties(name, newMsProperties) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] === name) {
					this.loopJsonCache["microServicePolicies"][policy]["configurationsJson"] = newMsProperties;
				}
			}
	}

	updateMicroServicePdpGroup(name, pdpGroup, pdpSubgroup) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] === name) {
					this.loopJsonCache["microServicePolicies"][policy]["pdpGroup"] = pdpGroup;
					this.loopJsonCache["microServicePolicies"][policy]["pdpSubgroup"] = pdpSubgroup;
				}
			}
	}

	updateGlobalProperties(newGlobalProperties) {
		this.loopJsonCache["globalPropertiesJson"] = newGlobalProperties;
	}

	updateOperationalPolicyProperties(name, newOpProperties) {
		for (var policy in this.loopJsonCache["operationalPolicies"]) {
				if (this.loopJsonCache["operationalPolicies"][policy]["name"] === name) {
					this.loopJsonCache["operationalPolicies"][policy]["configurationsJson"] = newOpProperties;
				}
			}
	}

	updateOperationalPolicyPdpGroup(name, pdpGroup, pdpSubgroup) {
		for (var policy in this.loopJsonCache["operationalPolicies"]) {
			if (this.loopJsonCache["operationalPolicies"][policy]["name"] === name) {
				this.loopJsonCache["operationalPolicies"][policy]["pdpGroup"] = pdpGroup;
				this.loopJsonCache["operationalPolicies"][policy]["pdpSubgroup"] = pdpSubgroup;
			}
		}
	}

	getLoopName() {
		return this.loopJsonCache["name"];
	}

	getOperationalPolicyJsonSchema() {
		return this.loopJsonCache["operationalPolicies"]["0"]["jsonRepresentation"];
	}

	getOperationalPolicies() {
		return this.loopJsonCache["operationalPolicies"];
	}

	getOperationalPoliciesNoJsonSchema() {
		var operationalPolicies = JSON.parse(JSON.stringify(this.loopJsonCache["operationalPolicies"]));
		delete operationalPolicies[0]["jsonRepresentation"];
		return operationalPolicies;
	}

	getGlobalProperties() {
		return this.loopJsonCache["globalPropertiesJson"];
	}

	getDcaeDeploymentProperties() {
		return this.loopJsonCache["globalPropertiesJson"]["dcaeDeployParameters"];
	}

	getMicroServicePolicies() {
		return this.loopJsonCache["microServicePolicies"];
	}

	getOperationalPolicyForName(name) {
		var opProperties=this.getOperationalPolicies();
		for (var policy in opProperties) {
			if (opProperties[policy]["name"] === name) {
				return opProperties[policy];
			}
		}
		return null;
	}

	getOperationalPolicyPropertiesForName(name) {
		var opConfig = this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["configurationsJson"];
		}
		return null;
	}

	getOperationalPolicyJsonRepresentationForName(name) {
		var opConfig = this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["jsonRepresentation"];
		}
		return null;
	}

	getOperationalPolicySupportedPdpGroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
		    if (opConfig["policyModel"]["policyPdpGroup"] !== undefined && opConfig["policyModel"]["policyPdpGroup"]["supportedPdpGroups"] !== undefined) {
			    return opConfig["policyModel"]["policyPdpGroup"]["supportedPdpGroups"];
			}
		}
		return [];
	}

	getOperationalPolicyPdpGroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["pdpGroup"];
		}
		return null;
	}

	getOperationalPolicyPdpSubgroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["pdpSubgroup"];
		}
		return null;
	}

	getMicroServiceSupportedPdpGroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
		    if (microService["policyModel"]["policyPdpGroup"] !== undefined && microService["policyModel"]["policyPdpGroup"]["supportedPdpGroups"] !== undefined) {
			    return microService["policyModel"]["policyPdpGroup"]["supportedPdpGroups"];
			}
		}
		return [];
	}

	getMicroServicePdpGroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
			return microService["pdpGroup"];
		}
		return null;
	}

	getMicroServicePdpSubgroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
			return microService["pdpSubgroup"];
		}
		return null;
	}

	getMicroServiceForName(name) {
		var msProperties=this.getMicroServicePolicies();
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] === name) {
				return msProperties[policy];
			}
		}
		return null;
	}

	getMicroServicePropertiesForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["configurationsJson"];
		}
		return null;
	}

	getMicroServiceJsonRepresentationForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["jsonRepresentation"];
		}
		return null;
	}

	getResourceDetailsVfProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VF"];
	}

	getResourceDetailsVfModuleProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VFModule"];
	}

	getLoopLogsArray() {
		return this.loopJsonCache.loopLogs;
	}

	getComputedState() {
		return this.loopJsonCache.lastComputedState;
	}

	getComponentStates() {
		return this.loopJsonCache.components;
	}

	getTemplateName() {
	    if (this.getLoopTemplate() !== undefined) {
        			return this.getLoopTemplate().name;
        }
	    return null;
	}

	getLoopTemplate() {
		return this.loopJsonCache["loopTemplate"];
	}

	isOpenLoopTemplate() {
		var loopTemplate = this.getLoopTemplate();
		if(loopTemplate != null && loopTemplate["allowedLoopType"] === "OPEN") {
			return true;
		}
		return false;
	}

	getAllLoopElementModels() {
	    var loopTemplate = this.getLoopTemplate();
	    var loopElementModels = [];
	    if(loopTemplate != null) {
	        for (var element of loopTemplate['loopElementModelsUsed']) {
	            loopElementModels.push(element['loopElementModel'])
	        }
	    }
	    return loopElementModels;
	}
}
