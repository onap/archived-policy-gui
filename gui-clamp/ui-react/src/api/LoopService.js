/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopService {
	static getLoopNames() {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/getAllNames', { method: 'GET', credentials: 'same-origin' })
			.then(function (response) {
				console.debug("GetLoopNames response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoopNames query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoopNames error received", error);
				return {};
			});
	}

	static createLoop(loopName, templateName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/create/' + loopName + '?templateName=' + templateName, {
			method: 'POST',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("CreateLoop response received: ", response.status);
				return response.json();
			})
			.catch(function (error) {
				console.error("CreateLoop error received", error);
				return "";
			});
	}

	static getLoop(loopName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/' + loopName, {
			method: 'GET',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("GetLoop response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoop query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoop error received", error);
				return {};
			});
	}

	static setMicroServiceProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateMicroservicePolicy/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateMicroservicePolicy response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateMicroservicePolicy query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateMicroservicePolicy error received", error);
				return "";
			});
	}
	
	static setOperationalPolicyProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateOperationalPolicies/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateOperationalPolicies response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateOperationalPolicies query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateOperationalPolicies error received", error);
				return "";
			});
	}

	static updateGlobalProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateGlobalProperties/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateGlobalProperties response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateGlobalProperties query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateGlobalProperties error received", error);
				return "";
			});
	}

	static refreshOperationalPolicyJson(loopName,operationalPolicyName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/refreshOperationalPolicyJsonSchema/' + loopName + '/' + operationalPolicyName, {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("Refresh Operational Policy Json Schema response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Refresh Operational Policy Json Schema query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Refresh Operational Policy Json Schema error received", error);
				return {};
			});
	}

		static refreshMicroServicePolicyJson(loopName,microServicePolicyName) {
    		return fetch(window.location.pathname + 'restservices/clds/v2/loop/refreshMicroServicePolicyJsonSchema/' + loopName + '/' + microServicePolicyName, {
    			method: 'PUT',
    			headers: {
    				"Content-Type": "application/json"
    			},
    			credentials: 'same-origin'
    		})
    			.then(function (response) {
    				console.debug("Refresh Operational Policy Json Schema response received: ", response.status);
    				if (response.ok) {
    					return response.json();
    				} else {
    					console.error("Refresh Operational Policy Json Schema query failed");
    					return {};
    				}
    			})
    			.catch(function (error) {
    				console.error("Refresh Operational Policy Json Schema error received", error);
    				return {};
    			});
    }

	static addOperationalPolicyType(loopName, policyType, policyVersion) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/addOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
				console.debug("Add Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					return response.text();
				}
			})
			.catch(function (error) {
				console.error("Add Operational Policy query failed");
				throw new Error(error);
			})
	}

	static removeOperationalPolicyType(loopName, policyType, policyVersion, policyName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/removeOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion + '/' + policyName , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
					console.debug("Remove Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Remove Operational Policy query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Remove Operational Policy error received", error);
				return {};
			});
	}
}
