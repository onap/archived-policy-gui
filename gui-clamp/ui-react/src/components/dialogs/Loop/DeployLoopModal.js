/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import LoopActionService from '../../../api/LoopActionService';
import LoopService from '../../../api/LoopService';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import Form from 'react-bootstrap/Form';
import Tabs from 'react-bootstrap/Tabs';
import Tab from 'react-bootstrap/Tab';
import styled from 'styled-components';
import Spinner from 'react-bootstrap/Spinner'

const StyledSpinnerDiv = styled.div`
	justify-content: center !important;
	display: flex !important;
`;

const ModalStyled = styled(Modal)`
	background-color: transparent;
`
const FormStyled = styled(Form.Group)`
	padding: .25rem 1.5rem;
`
export default class DeployLoopModal extends React.Component {


		
	constructor(props, context) {
		super(props, context);

		this.handleSave = this.handleSave.bind(this);
		this.handleClose = this.handleClose.bind(this);
		this.handleChange = this.handleChange.bind(this);
		this.refreshStatus = this.refreshStatus.bind(this);
		this.renderDeployParam = this.renderDeployParam.bind(this);
		this.renderSpinner = this.renderSpinner.bind(this);

		const propertiesJson = JSON.parse(JSON.stringify(this.props.loopCache.getGlobalProperties()));
		this.state = {
			loopCache: this.props.loopCache,
			temporaryPropertiesJson: propertiesJson,
			show: true,
			key: this.getInitialKeyValue(propertiesJson)
		};
	}
	getInitialKeyValue(temporaryPropertiesJson) {
		const deployJsonList = temporaryPropertiesJson["dcaeDeployParameters"];
		let initialKey;
		Object.keys(deployJsonList)
			.filter((obj) => Object.keys(deployJsonList).indexOf(obj) === 0)
			.map(obj =>
				initialKey = obj
		);
		return initialKey;
	}
	componentWillReceiveProps(newProps) {
		this.setState({
			loopName: newProps.loopCache.getLoopName(),
			show: true
		});
	}

	handleClose(){
		this.setState({ show: false });
		this.props.history.push('/');
	}

	renderSpinner() {
		if (this.state.deploying) {
			return (
				<StyledSpinnerDiv>
					<Spinner animation="border" role="status">
						<span className="sr-only">Loading...</span>
					</Spinner>
				</StyledSpinnerDiv>
			);
		} else {
			return (<div></div>);
		}
	}

	handleSave() {
		const loopName = this.props.loopCache.getLoopName();
		// save the global propserties
		this.setState({ deploying: true });
		LoopService.updateGlobalProperties(loopName, this.state.temporaryPropertiesJson).then(resp => {
			LoopActionService.performAction(loopName, "deploy").then(pars => {
				this.props.showSucAlert("Action deploy successfully performed");
				// refresh status and update loop logs
				this.refreshStatus(loopName);
			})
			.catch(error => {
				this.props.showFailAlert("Action deploy failed");
				// refresh status and update loop logs
				this.refreshStatus(loopName);
			});
		});
	}

	refreshStatus(loopName) {
		LoopActionService.refreshStatus(loopName).then(data => {
			this.props.updateLoopFunction(data);
			this.setState({ show: false, deploying: false });
			this.props.history.push('/');
		})
		.catch(error => {
			this.props.showFailAlert("Refresh status failed");
			this.setState({ show: false, deploying: false  });
			this.props.history.push('/');
		});
	}
	handleChange(event) {
		let deploymentParam = this.state.temporaryPropertiesJson["dcaeDeployParameters"];
		deploymentParam[this.state.key][event.target.name] = event.target.value;

		this.setState({temporaryPropertiesJson:{dcaeDeployParameters: deploymentParam}});
	}
	renderDeployParamTabs() {
		if (typeof (this.state.temporaryPropertiesJson) === "undefined") {
			 return "";
		}

		const deployJsonList = this.state.temporaryPropertiesJson["dcaeDeployParameters"];
		var indents = [];
		Object.keys(deployJsonList).map((item,key) =>
			indents.push(<Tab eventKey={item} title={item}>
				{this.renderDeployParam(deployJsonList[item])}
				</Tab>)
		);
		return indents;
	}
	renderDeployParam(deployJson) {
		var indents = [];
		Object.keys(deployJson).map((item,key) =>
		indents.push(<FormStyled>
				<Form.Label>{item}</Form.Label>
				<Form.Control type="text" name={item} onChange={this.handleChange} defaultValue={deployJson[item]}></Form.Control>
			</FormStyled>));
		return indents;
	}
	render() {
		return (
					<ModalStyled size="lg" show={this.state.show} onHide={this.handleClose} backdrop="static" keyboard={false} >
						<Modal.Header closeButton>
							<Modal.Title>Deployment parameters</Modal.Title>
						</Modal.Header>
						<Tabs id="controlled-tab-example" activeKey={this.state.key} onSelect={key => this.setState({ key })}>
						{this.renderDeployParamTabs()}
						</Tabs>
						{this.renderSpinner()}
						<Modal.Footer>
							<Button variant="secondary" type="null" onClick={this.handleClose}>Cancel</Button>
							<Button variant="primary" type="submit" onClick={this.handleSave}>Deploy</Button>
						</Modal.Footer>
					</ModalStyled>
		);
	}
}
