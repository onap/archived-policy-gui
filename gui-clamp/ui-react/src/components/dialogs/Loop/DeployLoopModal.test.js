/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import { shallow } from 'enzyme';
import DeployLoopModal from './DeployLoopModal';
import LoopCache from '../../../api/LoopCache';
import LoopActionService from '../../../api/LoopActionService';
import LoopService from '../../../api/LoopService';

describe('Verify DeployLoopModal', () => {
	const loopCache = new LoopCache({
		"name": "LOOP_Jbv1z_v1_0_ResourceInstanceName1_tca",
		"globalPropertiesJson": {
			"dcaeDeployParameters": {
				"testMs": {
					"location_id": "",
					"policy_id": "TCA_h2NMX_v1_0_ResourceInstanceName1_tca"
				}
			}
		}
	});

	it('Test the render method', () => {
		const component = shallow(
			<DeployLoopModal loopCache={loopCache}/>
		)

	expect(component).toMatchSnapshot();
	});
	
	it('Test handleClose', () => {
		const historyMock = { push: jest.fn() };
		const handleClose = jest.spyOn(DeployLoopModal.prototype,'handleClose');
		const component = shallow(<DeployLoopModal history={historyMock} loopCache={loopCache}/>)

		component.find('[variant="secondary"]').prop('onClick')();

		expect(handleClose).toHaveBeenCalledTimes(1);
		expect(historyMock.push.mock.calls[0]).toEqual([ '/']);
	});

	it('Test handleSave successful', async () => {
		const flushPromises = () => new Promise(setImmediate);
		const historyMock = { push: jest.fn() };
		const updateLoopFunction = jest.fn();
		const showSucAlert = jest.fn();
		const showFailAlert = jest.fn();
		const handleSave = jest.spyOn(DeployLoopModal.prototype,'handleSave');
		LoopService.updateGlobalProperties = jest.fn().mockImplementation(() => {
			return Promise.resolve({
				ok: true,
				status: 200,
				text: () => "OK"
			});
		});
		LoopActionService.performAction = jest.fn().mockImplementation(() => {
			return Promise.resolve({
				ok: true,
				status: 200,
				json: () => {}
			});
		});
		LoopActionService.refreshStatus = jest.fn().mockImplementation(() => {
			return Promise.resolve({
				ok: true,
				status: 200,
				json: () => {}
			});
		});

		const component = shallow(<DeployLoopModal history={historyMock} 
						loopCache={loopCache} updateLoopFunction={updateLoopFunction} showSucAlert={showSucAlert} showFailAlert={showFailAlert} />)

		component.find('[variant="primary"]').prop('onClick')();
		await flushPromises();
		component.update();

		expect(handleSave).toHaveBeenCalledTimes(1);
		expect(component.state('show')).toEqual(false);
		expect(historyMock.push.mock.calls[0]).toEqual([ '/']);
		handleSave.mockClear();
	});

	it('Onchange event', () => {
		const event = { target: { name: "location_id", value: "testLocation"} };
		const component = shallow(<DeployLoopModal loopCache={loopCache}/>);

		component.find('[name="location_id"]').simulate('change', event);
		component.update();
		expect(component.state('temporaryPropertiesJson').dcaeDeployParameters.testMs.location_id).toEqual("testLocation");
	});
});