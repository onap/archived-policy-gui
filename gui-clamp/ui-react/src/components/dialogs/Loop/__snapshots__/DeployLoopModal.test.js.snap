// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify DeployLoopModal Test the render method 1`] = `
<Styled(Modal)
  backdrop="static"
  keyboard={false}
  onHide={[Function]}
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      Deployment parameters
    </ModalTitle>
  </ModalHeader>
  <Tabs
    activeKey="testMs"
    id="controlled-tab-example"
    mountOnEnter={false}
    onSelect={[Function]}
    unmountOnExit={false}
    variant="tabs"
  >
    <Tab
      eventKey="testMs"
      key="testMs"
      title="testMs"
    >
      <Styled(FormGroup)
        key="location_id"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          location_id
        </FormLabel>
        <FormControl
          defaultValue=""
          name="location_id"
          onChange={[Function]}
          type="text"
        />
      </Styled(FormGroup)>
      <Styled(FormGroup)
        key="policy_id"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          policy_id
        </FormLabel>
        <FormControl
          defaultValue="TCA_h2NMX_v1_0_ResourceInstanceName1_tca"
          name="policy_id"
          onChange={[Function]}
          type="text"
        />
      </Styled(FormGroup)>
    </Tab>
  </Tabs>
  <div />
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="null"
      variant="secondary"
    >
      Cancel
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="submit"
      variant="primary"
    >
      Deploy
    </Button>
  </ModalFooter>
</Styled(Modal)>
`;
